clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -555,6 +555,9 @@ private void close(KafkaChannel channel, boolean processOutstanding) {
 
         channel.disconnect();
 
+        //avoid inconsistent connection states, see KAFKA-5417
+        connected.remove(channel.id());
```

Comment:
Yeah, I can't think of a strong reason for either option, so I guess we can leave it here.